File: typescript/tests/food_board.test.ts
// food_board.test.ts
// Verifies the CircleAI.Food port: recipes + ingredient search, meal logs,
// pantry usage + expiry.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { InMemoryFoodBoard, FoodDomainContext, recipe, mealLog, pantryItem } from "../src/food/index";

describe("InMemoryFoodBoard", () => {
  it("adds recipes and searches by ingredient (case-insensitive substring)", () => {
    const b = new InMemoryFoodBoard();
    b.addRecipe(recipe("r1", "Tomato Soup", ["Tomato", "Basil"], ["boil"], 4, 20));
    b.addRecipe(recipe("r2", "Omelette", ["Egg", "Cheese"], ["fry"], 1, 10));
    assert.equal(b.getRecipe("r1")?.title, "Tomato Soup");
    assert.deepEqual(
      b.searchByIngredient("tomato").map((r) => r.recipeId),
      ["r1"],
    );
    assert.equal(b.searchByIngredient("xyz").length, 0);
  });

  it("searchByIngredient throws on blank", () => {
    const b = new InMemoryFoodBoard();
    assert.throws(() => b.searchByIngredient("   "), /ingredient required/);
  });

  it("logs meals and returns them since a cutoff, oldest-first", () => {
    const b = new InMemoryFoodBoard();
    b.log(mealLog("m1", "u1", "r1", new Date("2026-01-01T12:00:00Z"), 1));
    b.log(mealLog("m2", "u1", "r2", new Date("2026-01-03T12:00:00Z"), 2));
    b.log(mealLog("m3", "u2", "r1", new Date("2026-01-03T12:00:00Z"), 1));
    assert.deepEqual(
      b.logsSince("u1", new Date("2026-01-02T00:00:00Z")).map((m) => m.logId),
      ["m2"],
    );
  });

  it("stocks the pantry, decrements on use (floored at 0), and lists positive items", () => {
    const b = new InMemoryFoodBoard();
    b.stockPantry(pantryItem("p1", "Flour", 1000, "g", null));
    b.stockPantry(pantryItem("p2", "Sugar", 500, "g", null));
    b.use("p1", 300);
    b.use("p2", 999); // over-use → floored to 0
    assert.equal(b.getRecipe("nope"), undefined);
    const remaining = b.pantry().map((p) => [p.pantryItemId, p.quantity]);
    assert.deepEqual(remaining, [["p1", 700]]);
  });

  it("use throws on unknown pantry item", () => {
    const b = new InMemoryFoodBoard();
    assert.throws(() => b.use("ghost", 1), /Unknown pantry item ghost/);
  });

  it("lists expiring items before a date, earliest-first", () => {
    const b = new InMemoryFoodBoard();
    b.stockPantry(pantryItem("p1", "Milk", 1, "l", new Date("2026-01-10T00:00:00Z")));
    b.stockPantry(pantryItem("p2", "Yoghurt", 1, "l", new Date("2026-01-05T00:00:00Z")));
    b.stockPantry(pantryItem("p3", "Salt", 1, "kg", null)); // no expiry
    assert.deepEqual(
      b.expiring(new Date("2026-01-08T00:00:00Z")).map((p) => p.pantryItemId),
      ["p2"],
    );
    assert.deepEqual(
      b.expiring(new Date("2026-01-20T00:00:00Z")).map((p) => p.pantryItemId),
      ["p2", "p1"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(FoodDomainContext.systemPromptSnippet.includes("[DOMAIN: Food]"));
    assert.deepEqual(FoodDomainContext.complianceFlags, ["Food_Safety_Act", "POPIA"]);
    assert.deepEqual(FoodDomainContext.suggestedTools, ["recipe_tools", "nutrition_db", "shopping_list", "web_search"]);
  });
});

File: go/tests/food_board_test.go
// food_board_test.go
//
// Verifies the CircleAI.Food port (food_board.go): recipe add/get, ingredient
// search (case-insensitive), meal-log since, and pantry stock/use/expiring.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestFood_RecipesAndSearch(t *testing.T) {
	b := circleai.NewInMemoryFoodBoard()
	b.AddRecipe(circleai.Recipe{RecipeId: "r1", Title: "Pasta", Ingredients: []string{"Tomato", "Basil"}, Servings: 2, PrepMinutes: 20})
	b.AddRecipe(circleai.Recipe{RecipeId: "r2", Title: "Salad", Ingredients: []string{"Lettuce", "tomato"}, Servings: 1, PrepMinutes: 10})
	if got, ok := b.GetRecipe("r1"); !ok || got.Title != "Pasta" {
		t.Fatalf("get recipe = %+v ok=%v", got, ok)
	}
	hits := b.SearchByIngredient("TOMATO")
	if len(hits) != 2 || hits[0].RecipeId != "r1" || hits[1].RecipeId != "r2" {
		t.Fatalf("ingredient search failed: %+v", hits)
	}
}

func TestFood_PantryAndExpiring(t *testing.T) {
	b := circleai.NewInMemoryFoodBoard()
	early := time.Date(2026, 1, 10, 0, 0, 0, 0, time.UTC)
	late := time.Date(2026, 3, 1, 0, 0, 0, 0, time.UTC)
	b.StockPantry(circleai.PantryItem{PantryItemId: "p1", Name: "Milk", Quantity: 2, Unit: "L", BestBefore: &early})
	b.StockPantry(circleai.PantryItem{PantryItemId: "p2", Name: "Rice", Quantity: 5, Unit: "kg", BestBefore: &late})
	b.StockPantry(circleai.PantryItem{PantryItemId: "p3", Name: "Salt", Quantity: 1, Unit: "kg"})

	if err := b.Use("p1", 3); err != nil {
		t.Fatalf("use: %v", err)
	}
	// p1 floored to 0 -> excluded from Pantry.
	pantry := b.Pantry()
	if len(pantry) != 2 || pantry[0].PantryItemId != "p2" || pantry[1].PantryItemId != "p3" {
		t.Fatalf("pantry (qty>0, sorted) failed: %+v", pantry)
	}
	if err := b.Use("ghost", 1); err == nil {
		t.Fatalf("use unknown item must error")
	}

	exp := b.Expiring(time.Date(2026, 2, 1, 0, 0, 0, 0, time.UTC))
	if len(exp) != 1 || exp[0].PantryItemId != "p1" {
		t.Fatalf("expiring failed: %+v", exp)
	}

	// Meal logs since.
	now := time.Date(2026, 7, 8, 12, 0, 0, 0, time.UTC)
	b.Log(circleai.MealLog{LogId: "m1", UserId: "u1", RecipeId: "r1", AtUtc: now, Servings: 1})
	b.Log(circleai.MealLog{LogId: "m2", UserId: "u1", RecipeId: "r2", AtUtc: now.Add(-72 * time.Hour), Servings: 2})
	logs := b.LogsSince("u1", now.Add(-48*time.Hour))
	if len(logs) != 1 || logs[0].LogId != "m1" {
		t.Fatalf("logs-since failed: %+v", logs)
	}
}
